### ISSUE TITLE
Self-closing component tags (<my-comp/> instead of <my-comp></my-comp>)?

### ISSUE THREAD

**Issue by samclaus** [angular/angular#39525]
# 🚀 feature request

### Relevant Package

This feature request is for `@angular/compiler`? Unsure.

### Description

The issue title pretty much covers it but here is another example:

```HTML
<super-duper-long-component-name [prop]="someVar" />
<!-- instead of -->
<super-duper-long-component-name [prop]="someVar"></super-duper-long-component-name>
```

There is an existing issue for this, #673, but it was closed due to inactivity.

That issue got half-rejected (it was not resolved) on the basis of the HTML spec, which does not allow self-closing tags. That makes a total of **zero** sense to me. The HTML template language is already far beyond what vanilla HTML offers, and that is the entire point of using a framework like Angular. Many other major frameworks (React/Vue/Svelte to name a few) already support self-closing component tags.

I understand that Angular templates were being parsed in the browser before, but why should ease-of-life syntax sugar be banned when using AOT compilation? Sure, some people still use JIT-compiled Angular, but then again, some people still use Internet Explorer in 2020. If they want to suffer so be it but why refuse to add excellent features for the rest of us? There are plenty of antifeatures out there but I personally think self-closing tags are pretty intuitive and improve the code quality significantly; from the looks of #673 it seems like a lot of other people agree.

Please reconsider self-closing component tags.

### Describe alternatives you've considered

Continue to sigh as I write needlessly verbose code and write checks for two different nullish types.

**Comment by spock123** [comment]
This has been posted over and over again.

Afaik self closing tags are not valid HTML tags.

**Comment by Airblader** [comment]
> The HTML template language is already far beyond what vanilla HTML offers

The Angular template syntax is valid HTML5.

**Comment by vytautas-petrikas** [comment]
How would content projection work? :?

**Comment by Airblader** [comment]
@vytautas-petrikas That wouldn't change, in this case you'd have to write out the closing tag as you do now. Self-closing tags are just intended as syntactical sugar when there is no content inside.

**Comment by alexzuza** [comment]
Duplicate of https://github.com/angular/angular/issues/673

**Comment by spock123** [comment]
> Duplicate of #673

I think @samterainsights already mentioned that he wanted to bring it to attention again (given the old issue has been closed for a long time).

The issue is that many developers don't understand the need to adhere to strict html5, when all other frameworks value developer experience higher. There should be no technical issue in supporting it, so I'm not sure I fully understand the religious adherence to html5 specs.

Maybe someone smarter than me can explain the value of strictly following html5 specs, when it apparently is not an issue for vue, react, svelte and what not. Cheers

**Comment by AleksanderBodurri** [comment]
Isn't JIT still used in non-ivy apps for development? There are still a lot of apps that have not made the change to Ivy from View Engine (even if they have upgraded to Angular 9/10).

**Comment by samterainsights** [comment]
> Isn't JIT still used in non-ivy apps for development? There are still a lot of apps that have not made the change to Ivy from View Engine (even if they have upgraded to Angular 9/10).

Yes, many people still use JIT but you do not need Ivy for AOT compilation--it has been supported since I switched from AngularJS to Angular 5 (and maybe earlier). If you try to JIT compile templates using the syntax sugar, you will get an error. So if you want to JIT compile for some reason, simply don't use the feature. At the time of writing, Angular AOT compilation is painfully slow even with Ivy, so I understand why people still opt to use JIT during development. My hope is that the Angular team will improve the compiler performance significantly over the next couple of years (maybe rewrite in a more performant language like Rust--that is already happening for TypeScript thanks to the DenoJS project), and then the JIT compilation will be irrelevant and everyone can enjoy the benefits of clean syntax sugar.

**Comment by sod** [comment]
> The Angular template syntax is valid HTML5.

You sure about that? HTML ist mostly case insensitive where as angular markup is case sensitive.

**Comment by spock123** [comment]
> > The Angular template syntax is valid HTML5.
> 
> You sure about that? HTML ist mostly case insensitive where as angular markup is case sensitive.

I think you need to look at your logic again.

That doesn't mean template syntax is not valid HTML5 then (because as you said, HTML5 is not case sensitive).

**Comment by lazarljubenovic** [comment]
> > The HTML template language is already far beyond what vanilla HTML offers
> 
> The Angular template syntax is valid HTML5.

It all depends on how you define "valid HTML5". There's a syntax level, and a semantic level: Is `<div href="oops"></div>` valid? Well, _kinda_. The syntax is fine, but `href` is not a valid `div` attribute. Browsers will swallow this and HTML5 spec was updated to just ignore nonsense attributes because of historical reasons and backwards compatibility. Any IDE will scream at you for attempting to href a div, and for all practical reasons people will say "hey that's not valid". 

Secondly, a "valid HTML5" isn't even a complete phrase. A valid HTML5 what? A document? Angular templates are not HTML5 documents, they are just chunks (no doctype at the top, no `<html>`). 

That said, custom elements DO require a closing tag according to HTML5 spec: https://html.spec.whatwg.org/multipage/syntax.html#syntax-elements Whenever we _can_, we _should_ align with the custom elements spec. The tokenizer/parser mixture already needs information about which tags are self-closing in order to properly parse the document, so it would have to end up as `<my-el></my-el>` in the final DOM anyway.

Even without all this, I doubt this would be implemented in next 10 years. It's been ages since the last update on what the templates are capable of. The JS part of the template syntax has been stuck on ES3 since "Angular 2" Alpha/Beta releases. You can't write a lambda function in the template, you can't use rest/spread operator, you can't use hexadecimal literals, and don't even get me started on 2020 additions of  `?.` and `??`.

So my suggestion is to get used to writing both the opening and closing tag. You'll carry the good habit into HTML5's custom elements. Besides, any IDE capable of understand HTML is going to add the closing tag for you automatically anyway.

**Comment by lazarljubenovic** [comment]
Another thing.

> Many other major frameworks (React/Vue/Svelte to name a few) already support self-closing component tags.

> Please reconsider self-closing component tags.

This is because in these frameworks, a component is defined by a tag name and there's a clear one-to-one mapping. In Angular, a component has a selector. The template compiler goes through the template and does a check for each present element to see if it matches with a selector of any component included in the module where the component whose template is being checked is declared. You can have a component with a selector `div#id.class[random-attr]`.

In Angular, you don't "create a component with a tag my-cmp". You create a component with a selector "my-cmp", and when a template contains an element which matches it (which happens to be a custom element since there's no `my-cmp` in HTML5), the template of the matched component gets splashed inside (obviously there's more going on in this step but this suffices for this discussion). So the issue with "allow this for _components_" is that you don't know if it's a component before it ends up being a component -- at which point, your template already has to be valid, and therefore the tag must be closed (because a custom element's tag must be closed).

React doesn't even have a concept of tags. You're probably thinking of JSX, which is just syntax sugar for calling functions with a special signature. Instead of doing `h('x', { foo: 'bar', children: [h('y')] })`, you can write `<x foo="bar"><y/></x>`. When the second parameter doesn't have a `children` prop, you can self-close the tag. That's all. There's no HTML there, it's just a tiny transpiler on top of ES/TS which allows you to fool yourself into believing you're writing HTML (and then you get weird habits of doing `<input/>` which should really be `<input>` since it's a void tag and not self-closing tag).

**Comment by samterainsights** [comment]
> This is because in these frameworks, a component is defined by a tag name and there's a clear one-to-one mapping. In Angular, a component has a selector. The template compiler goes through the template and does a check for each present element to see if it matches with a selector of any component included in the module where the component whose template is being checked is declared. You can have a component with a selector `div#id.class[random-attr]`.

I know the selectors are not necessarily tag names. That information does not invalidate my issue whatsoever. Omitting the closing tag in favor of a slash at the end of the opening tag breaks HTML5 syntax, but it is nothing more than first-pass syntax sugar for the compiler. Angular templates are parsed into JavaScript and do not get run in any manner by the browser. The only technical issue, if I heard correctly, is that when templates are compiled JIT (opposed to AOT), the Angular runtime uses a browser API to parse the templates, and that parser is only capable of parsing valid HTML and returns incorrect AST information if a self-closing tag is used. Even for this case, Angular could use a custom parser written in JavaScript, which is what I assume they already do for the NodeJS-based Angular compiler.

Technically speaking, self-closing tags in Angular are very feasible. I think spock123 summed up my viewpoint very well in his comment:

> Maybe someone smarter than me can explain the value of strictly following html5 specs, when it apparently is not an issue for vue, react, svelte and what not.

I suppose that is the more general question here.

**Comment by iRealNirmal** [comment]
I would also like to see this feature, I also did some code change and it's quite easy fix (although I would have to add some good amount of test case this change has to be done in compiler package). Earlier this section coding was done by @vicb, so I would like to hear input from him as well before raising PR.

@petebacondarwin @AndrewKushnir if you have any inputs  then can you also share ?

**Comment by petebacondarwin** [comment]
As @lazarljubenovic said:

> That said, custom elements DO require a closing tag according to HTML5 spec: https://html.spec.whatwg.org/multipage/syntax.html#syntax-elements Whenever we can, we should align with the custom elements spec. The tokenizer/parser mixture already needs information about which tags are self-closing in order to properly parse the document, so it would have to end up as <my-el></my-el> in the final DOM anyway.

**Comment by iRealNirmal** [comment]
understood thanks

**Comment by spock123** [comment]
One thing is adhering to HTML specs -but could the compiler not take care of this?

This way we get best of both worlds - our developer experience is improved a lot, and the end result (compilation result) is following HTML spec.

**Comment by petebacondarwin** [comment]
@spock123 - when you say "a lot" how do you measure that? Number of key presses requires by the developer is unlikely to be effected since IDEs will provide autocompletion of the closing tag. What are the other benefits to the developer?

**Comment by samclaus** [comment]
@petebacondarwin I think you're missing the point. We're already talking about first-world issues here. I'm not trying to be rude. You are absolutely correct that, yes, it's already _good enough_. Nitpicky issues like this one are about striving to perfect the framework. I use Angular and do not contribute to it. If you wish me to f**k off, I will do so courteously. The thing that bothers me is this religious adherence to the HTML spec when the whole point of component frameworks is to escape the limitations of HTML. None of these client-side (so not including the NodeJS SSR support) frameworks function without JavaScript. They do not compile to HTML. They compile to JavaScript. The browser does not read Angular templates. Angular templates should follow the HTML spec because people often want to step outside the framework and work with raw DOM elements so Angular should not water down any DOM functionality. A component host is just a plain old `HTMLElement` meaning we can write vanilla JS to style it, remove it, whatever. It also means people don't need to learn new attributes and property names because writing templates is _like_ writing regular HTML. Those are the benefits of templates _looking_ like HTML. The benefits of adding syntax sugar like closing tag shorthand is that it makes the code considerably nicer to look at and thus makes new and even familiar codebases much more approachable. What benefit does **not** adding this syntax sugar yield?

I have an extremely deep appreciation for feature gating. I think people should fight on every step of software design, because otherwise we end up with bad software (there is a LOT of bad software). In this case, I am a user of Angular fighting with the developers of Angular. Angular developers should feature gate because otherwise the framework ends up cluttered with a bunch of antifeatures. That said, syntax sugar is typically not heavy on the brain to think about and makes coding more fun. Lastly, Angular already failed in the lightweight/elegant category: the documentation is huge and required multiple days for me to digest completely. It's more heavily opinionated than Svelte, Vue, and React. As such, you would at least expect it to be pleasant to write since everything is already thought out for you. That has not been the case for me. I have found Svelte to be--yes this is cliché but it expresses accurately what I feel--a breath of fresh air. Component templates are just way shorter, way faster to read, and way faster to write. Svelte favored solid foundations with few assumptions and then plenty of syntax sugar to aid the developer experience over a heavily opinionated dev stack. I use Angular at work so I have not been able to experiment with Svelte as much as I want and I can already foresee a couple of pain points that Angular has solved for me (like injecting parent components for inter-component coordination, which is incredibly useful for keyboard navigation directives). However, I might even flesh out my own intercomponent-communication primitives and end up liking them more. I'm getting off track but basically my point is this: feature gating is good but Angular (and especially Angular Material!) already failed to feature gate and right now you are dismissing a feature without any reasoning beyond "it's not valid HTML" and no explanation as to why following the HTML spec is so d**n important. Fin.

**Comment by IgorMinar** [comment]
@samclaus  there is no need to "fight" with us 😄 We are all interested in improving Angular and making it easier to use, and we'd like to do so via a *respectful and collaborative discussion*.

In this particular case, there are many nuances to simply allowing the nicer and shorter syntax, especially around what this syntax means for corner-cases (e.g. components with an element selector without a dash (e.g. `<mycomponent>`, or components that use valid html element selectors e.g. `button` but this native element is not a [void element](https://html.spec.whatwg.org/multipage/syntax.html#void-elements)). There are also additional considerations around template editing, syntax highlighting, autocompletion, linting etc - all of these tools nowadays work with Angular because they already support standard html. Once we move away from the spec, there is a possibility that common developer workflows could be affected or outright broken.

In short, we are all very much interested in this feature, but before we can go ahead and "just add it", we'll need to do proper due diligence and testing to understand the full impact of such change.

Since this whole issue is so nuanced, and even we are not sure about the full scope of things that might depend on us following the html spec in this area, I expect that we'll create a full RFC for this to collect the community feedback, and hear from developers that have built their custom tooling with the assumption that Angular templates are HTML compliant.

**Comment by samclaus** [comment]
@IgorMinar That was a well-explained response. No one mentioned things like tooling expecting HTML-compliant syntax and it didn't cross my mind since I was only thinking at the scope of the framework. Thanks.

**Comment by spock123** [comment]
Thank you so much guys, this is awesome!

**Comment by karol-depka** [comment]
Hi. What is the status of this in 2023? Did the PR go in? If so, in which Angular version will it be available? Thanks.

**Comment by crisbeto** [comment]
It's in 15.1.0.
